repository_pool.go

```diff
@@ -0,0 +1,234 @@
+package gitquery
+
+import (
+	"io"
+	"runtime"
+	"sync"
+
+	"gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/go-mysql-server.v0/sql"
+)
+
+// Repository struct holds an initialized repository and its ID
+type Repository struct {
+	ID   string
+	Repo *git.Repository
+}
+
+// NewRepository creates and initializes a new Repository structure
+func NewRepository(id string, repo *git.Repository) Repository {
+	return Repository{
+		ID:   id,
+		Repo: repo,
+	}
+}
+
+// RepositoryPool holds a pool of initialized git repositories and
+// functionality to open and iterate them.
+type RepositoryPool struct {
+	repositories []Repository
+}
+
+// NewRepositoryPool initializes a new RepositoryPool
+func NewRepositoryPool() RepositoryPool {
+	return RepositoryPool{}
+}
+
+// Add inserts a new repository in the pool
+func (p *RepositoryPool) Add(id string, repo *git.Repository) {
+	repository := NewRepository(id, repo)
+	p.repositories = append(p.repositories, repository)
+}
+
+// AddGit opens a new git repository and adds it to the pool. It
+// also sets its path as ID.
+func (p *RepositoryPool) AddGit(path string) (string, error) {
+	repo, err := git.PlainOpen(path)
+	if err != nil {
+		return "", err
+	}
+
+	p.Add(path, repo)
+
+	return path, nil
+}
+
+// GetPos retrieves a repository at a given position. It returns false
+// as second return value if the position is out of bounds.
+func (p *RepositoryPool) GetPos(pos int) (*Repository, bool) {
+	if pos >= len(p.repositories) {
+		return nil, false
+	}
+
+	return &p.repositories[pos], true
+}
+
+// RepoIter creates a new Repository iterator
+func (p *RepositoryPool) RepoIter() (*RepositoryIter, error) {
+	iter := &RepositoryIter{
+		pos:  0,
+		pool: p,
+	}
+
+	return iter, nil
+}
+
+// RepositoryIter iterates over all repositories in the pool
+type RepositoryIter struct {
+	pos  int
+	pool *RepositoryPool
+}
+
+// Next retrieves the next Repository. It returns io.EOF as error
+// when there are no more Repositories to retrieve.
+func (i *RepositoryIter) Next() (*Repository, error) {
+	r, ok := i.pool.GetPos(i.pos)
+	if !ok {
+		return nil, io.EOF
+	}
+
+	i.pos++
+
+	return r, nil
+}
+
+// Close finished iterator. It's no-op.
+func (i *RepositoryIter) Close() error {
+	return nil
+}
+
+// RowRepoIterImplementation is the interface needed by each iterator
+// implementation
+type RowRepoIterImplementation interface {
```

Comment:
This is an interface, so it's a bit weird it has the `Implementation` suffix. Can we name it something else?

Reply:
I see it's because we have a `RowRepoIter` struct. We have a few options, though:

- Call the struct `BaseRowRepoIter` or `DefaultRowRepoIter`
- Just make it private as `rowRepoIter` if it does not need to be accessed from outside and let the constructor `NewRowRepoIter` return the interface type instead.

Reply:
How about renaming the interface `rowRepoIterable`? It's the struct which needs to be returned. The interface is used internally by the `RowRepoIter` to retrieve rows from each repo.

Reply:
But don't the `NewIterator` methods return `RowRepoIterImplementation` right now?